src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -234,15 +234,18 @@ IEnumerator IEnumerable.GetEnumerator()
         // InvalidOperationException.
         public T Dequeue()
         {
-            if (_size == 0)
+            int head = _head;
+            T[] array = _array;
+            
+            if (_size == 0 || (uint)head >= (uint)array.Length)
```

Comment:
Why is the latter condition necessary?

Reply:
To enable the range check elimination on the following `array[head]`. The real win is on reference types, where this access is two times.

Reply:
Isn't it strange that we can do the check in this manner more efficiently than the JIT can?  What prevents the JIT from doing its bounds check in a similar manner?

Reply:
The JIT has no way of knowing that `head` is a valid array index so it generates normal bounds checks for `array[head]`, that's something like  `if ((uint)head >= (uint)array.Length) throw IndexOutOfRangeException();`.

If we do the check manually we can piggy back on the existing throw instead of generating a separate `throw IndexOutOfRangeException()`.

It's a rather creative use of the `(uint)head >= (uint)array.Length` trick. Not entirely sure it's worth it.

Reply:
@gfoidl, does it actually improve throughput, or just decrease the asm size?

Reply:
It improves throughput, as shown by the numbers in https://github.com/dotnet/corefx/pull/26087#issue-284945181, due to the elimination of the range check that is saved on the second access -- although only for rerfence types.

Until now I didn't check the asm size, but it goes down from 177 bytes to 163 bytes.

Reply:
> It improves throughput, as shown by the numbers 

Don't those numbers include all of the changes made in this PR, not just the addition of that extra condition?

Reply:
You are right, I forgot about the change in `MoveNext` that also has an influence to the numbers. So I'll do some benchmark that focuses on this.

Reply:
``` ini

BenchmarkDotNet=v0.10.11, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.125)
Processor=Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), ProcessorCount=8
Frequency=2742191 Hz, Resolution=364.6719 ns, Timer=TSC
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
|          Method |      Mean |     Error |    StdDev | Scaled | ScaledSD |
|---------------- |----------:|----------:|----------:|-------:|---------:|
|      Dequeue_PR | 0.7148 ns | 0.0234 ns | 0.0195 ns |   1.00 |     0.00 |
|  Dequeue_wo_RCE | 0.7514 ns | 0.0791 ns | 0.0740 ns |   1.05 |     0.10 |

`Dequeue_PR` is the method from this PR.
`Dequeue_wo_RCE` is:
```c#
public T Dequeue()
{
    if (_size == 0)
    {
        ThrowForEmptyQueue();
    }

    T removed = _array[_head];
    if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
    {
        _array[_head] = default;
    }
    MoveNext(ref _head);
    _size--;
    _version++;
    return removed;
}
```

Reply:
What if you use the code exactly as it is in your PR, and just remove " || (uint)head >= (uint)array.Length"?

Reply:
The improvement here most likely comes from caching `_array` and `_head` in local variables, not from the manual bounds check.

Reply:
> What if you use the code exactly as it is in your PR

Then it is equal in perf (within noise, so I can't say which one is faster). 

The dasm is exactely as @mikedn pointed out in https://github.com/dotnet/corefx/pull/26087#discussion_r164482687

With the manual bounds check
```asm
G_M11457_IG02:
       448B7318             mov      r14d, dword ptr [rbx+24]
       4C8B7B08             mov      r15, gword ptr [rbx+8]
       837B2000             cmp      dword ptr [rbx+32], 0
       7474                 je       SHORT G_M11457_IG08
       418B7F08             mov      edi, dword ptr [r15+8]
       413BFE               cmp      edi, r14d
       766B                 jbe      SHORT G_M11457_IG08

; more code 

G_M11457_IG08:
       488BFB               mov      rdi, rbx
       E8D6F6FFFF           call     ...:ThrowForEmptyQueue():this
       CC                   int3    
```

Without the manual bounds check:
```asm
G_M7328_IG02:
       448B7318             mov      r14d, dword ptr [rbx+24]
       4C8B7B08             mov      r15, gword ptr [rbx+8]
       837B2000             cmp      dword ptr [rbx+32], 0
       7471                 je       SHORT G_M7328_IG08

G_M7328_IG03:
       453B7708             cmp      r14d, dword ptr [r15+8]
       7373                 jae      SHORT G_M7328_IG09

; more code

G_M7328_IG08:
       488BFB               mov      rdi, rbx
       E859F7FFFF           call     ...:ThrowForEmptyQueue():this

G_M7328_IG09:
       E88C379878           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
```

It is cool that the JIT does emit only one bounds check for 
```c#
T removed = array[head];
if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
{
	array[head] = default;
}
```
```asm
G_M7328_IG04:
       488BF8               mov      rdi, rax
       E8222965FF           call     System.Runtime.CompilerServices.RuntimeHelpers:IsReferenceOrContainsReferences():bool
       85C0                 test     eax, eax
       740A                 je       SHORT G_M7328_IG05
       4963C6               movsxd   rax, r14d
       33FF                 xor      rdi, rdi
       49897CC710           mov      gword ptr [r15+8*rax+16], rdi    ; no bounds check 
```
I didn't know about that fact, and have to admit that I didn't investigate this case thoroughly enough, hence the manual check in this PR.

So in conclusion there is no real benefit from this manual check -- except the saving of 2 bytes of code. But the code is too big for inlining anyway, therefore this should not matter.

I will update the PR to remove the manual bound check.